File: Example-Mac/PINRemoteImage/ProcessingViewController.m
//
//  ProcessingViewController.m
//  PINRemoteImage
//
//  Created by Michael Schneider on 1/6/16.
//  Copyright © 2016 mischneider. All rights reserved.
//

#import "ProcessingViewController.h"

#import <PINRemoteImage/PINImageView+PINRemoteImage.h>
#import <PINRemoteImage/PINImage+DecodedImage.h>

@interface ProcessingViewController ()
@property (weak) IBOutlet NSImageView *imageView;

@end

@implementation ProcessingViewController

#pragma mark - Lifecycle

- (instancetype)init
{
    self = [super initWithNibName:NSStringFromClass(self.class) bundle:nil];
    if (self == nil) { return self; }
    return self;
}


#pragma mark - NSViewController

- (void)viewWillAppear
{
    [super viewWillAppear];
    
    [self.imageView pin_setImageFromURL:[NSURL URLWithString:@"https://s-media-cache-ak0.pinimg.com/736x/5b/c6/c5/5bc6c5387ff6f104fd642f2b375efba3.jpg"] processorKey:@"rounded" processor:^NSImage *(PINRemoteImageManagerResult *result, NSUInteger *cost)
     {
         NSImage *image = result.image;
         CGFloat radius = 7.0f;
         CGSize targetSize = CGSizeMake(200, 300);
         CGRect imageRect = CGRectMake(0, 0, targetSize.width, targetSize.height);

         NSImage *processedImage = [[NSImage alloc] initWithSize:targetSize];
         [processedImage lockFocus];
         
         NSBezierPath *bezierPath = [NSBezierPath bezierPathWithRoundedRect:imageRect xRadius:radius yRadius:radius];
         [bezierPath addClip];
         
         CGFloat widthMultiplier = targetSize.width / image.size.width;
         CGFloat heightMultiplier = targetSize.height / image.size.height;
         CGFloat sizeMultiplier = MAX(widthMultiplier, heightMultiplier);
         
         CGRect drawRect = CGRectMake(0, 0, image.size.width * sizeMultiplier, image.size.height * sizeMultiplier);
         if (CGRectGetMaxX(drawRect) > CGRectGetMaxX(imageRect)) {
             drawRect.origin.x -= (CGRectGetMaxX(drawRect) - CGRectGetMaxX(imageRect)) / 2.0;
         }
         if (CGRectGetMaxY(drawRect) > CGRectGetMaxY(imageRect)) {
             drawRect.origin.y -= (CGRectGetMaxY(drawRect) - CGRectGetMaxY(imageRect)) / 2.0;
         }
         
         [image drawInRect:drawRect];
         
         [[NSColor redColor] setStroke];
         [bezierPath setLineWidth:5.0];
         [bezierPath stroke];
        
         NSImage *logo = [NSImage imageNamed:@"white-pinterest-logo"];
         CGContextRef ctx = [[NSGraphicsContext currentContext] graphicsPort];
         CGContextSetBlendMode(ctx, kCGBlendModeOverlay);
         CGContextSetAlpha(ctx, 0.5);
         CGContextDrawImage(ctx, CGRectMake(90, 10, logo.size.width, logo.size.height), [logo CGImage]);
         
         [processedImage unlockFocus];
         
         return processedImage;
     }];
}

@end

File: Swift-Example/Swift-Example/ProcessingViewController.swift
//
//  ProcessingViewController.swift
//  Swift-Example
//
//  Created by Marius Landwehr on 23.01.16.
//  Copyright © 2016 Marius Landwehr. All rights reserved.
//

import UIKit
import PINRemoteImage

class ProcessingViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        let url = URL(string: "https://s-media-cache-ak0.pinimg.com/736x/5b/c6/c5/5bc6c5387ff6f104fd642f2b375efba3.jpg")
        
        imageView.pin_setImage(from: url, processorKey: "rounded") { (result, unsafePointer) -> UIImage? in
            
            guard let image = result.image else { return nil }
            
            let radius : CGFloat = 7.0
            let targetSize = CGSize(width: 200, height: 300)
            let imageRect = CGRect(x: 0, y: 0, width: targetSize.width, height: targetSize.height)
            
            UIGraphicsBeginImageContext(imageRect.size)
            
            let bezierPath = UIBezierPath(roundedRect: imageRect, cornerRadius: radius)
            bezierPath.addClip()
            
            let widthMultiplier : CGFloat = targetSize.width / image.size.width
            let heightMultiplier : CGFloat = targetSize.height / image.size.height
            let sizeMultiplier = max(widthMultiplier, heightMultiplier)
            
            var drawRect = CGRect(x: 0, y: 0, width: image.size.width * sizeMultiplier, height: image.size.height * sizeMultiplier)
            if (drawRect.maxX > imageRect.maxX) {
                drawRect.origin.x -= (drawRect.maxX - imageRect.maxX) / 2
            }
            if (drawRect.maxY > imageRect.maxY) {
                drawRect.origin.y -= (drawRect.maxY - imageRect.maxY) / 2
            }
            
            image.draw(in: drawRect)
            
            UIColor.red.setStroke()
            bezierPath.lineWidth = 5.0
            bezierPath.stroke()
            
            let ctx = UIGraphicsGetCurrentContext()
            ctx?.setBlendMode(CGBlendMode.overlay)
            ctx?.setAlpha(0.5)
            
            let logo = UIImage(named: "white-pinterest-logo")
            ctx?.scaleBy(x: 1.0, y: -1.0)
            ctx?.translateBy(x: 0.0, y: -drawRect.size.height)
            
            if let coreGraphicsImage = logo?.cgImage {
                ctx?.draw(coreGraphicsImage, in: CGRect(x: 90, y: 10, width: logo!.size.width, height: logo!.size.height))
            }
            
            let processedImage = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()
            
            return processedImage
            
        }
    }
}
